File: tunnel-server/src/main/java/com/tuuzed/tunnel/server/TunnelServerChannelHandler.java
package com.tuuzed.tunnel.server;

import com.tuuzed.tunnel.interceptor.ProtoRequestInterceptor;
import com.tuuzed.tunnel.proto.Proto;
import com.tuuzed.tunnel.proto.ProtoException;
import com.tuuzed.tunnel.proto.ProtoMessage;
import com.tuuzed.tunnel.proto.ProtoRequest;
import com.tuuzed.tunnel.server.http.HttpTunnelServer;
import com.tuuzed.tunnel.server.internal.AttributeKeys;
import com.tuuzed.tunnel.server.internal.ServerTunnelSessions;
import com.tuuzed.tunnel.server.internal.TokenProducer;
import com.tuuzed.tunnel.server.tcp.TcpTunnelServer;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;

@SuppressWarnings("Duplicates")
public class TunnelServerChannelHandler extends SimpleChannelInboundHandler<ProtoMessage> {
    private static final Logger logger = LoggerFactory.getLogger(TunnelServerChannelHandler.class);
    @NotNull
    private final TcpTunnelServer tcpTunnelServer;
    @Nullable
    private final HttpTunnelServer httpTunnelServer;
    @Nullable
    private final HttpTunnelServer httpsTunnelServer;
    @NotNull
    private final ProtoRequestInterceptor protoRequestInterceptor;
    @NotNull
    private final TokenProducer tunnelTokenProducer;

    public TunnelServerChannelHandler(
        @NotNull TcpTunnelServer tcpTunnelServer,
        @Nullable HttpTunnelServer httpTunnelServer,
        @Nullable HttpTunnelServer httpsTunnelServer,
        @NotNull ProtoRequestInterceptor protoRequestInterceptor,
        @NotNull TokenProducer tunnelTokenProducer
    ) {
        this.tcpTunnelServer = tcpTunnelServer;
        this.httpTunnelServer = httpTunnelServer;
        this.httpsTunnelServer = httpsTunnelServer;
        this.protoRequestInterceptor = protoRequestInterceptor;
        this.tunnelTokenProducer = tunnelTokenProducer;
    }


    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        /*
         * 隧道连接断开
         */
        logger.trace("channelInactive: {}", ctx);
        final ServerTunnelSessions tunnelSessions = ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).get();
        if (tunnelSessions != null) {
            final Proto proto = tunnelSessions.protoRequest().proto();
            switch (proto) {
                case TCP:
                    final long tunnelToken = tunnelSessions.tunnelToken();
                    tcpTunnelServer.shutdownTunnel(tunnelToken);
                    break;
                case HTTP:
                    if (httpTunnelServer != null) {
                        httpTunnelServer.unregister(tunnelSessions.protoRequest().vhost());
                    }
                    break;
                case HTTPS:
                    if (httpsTunnelServer != null) {
                        httpsTunnelServer.unregister(tunnelSessions.protoRequest().vhost());
                    }
                    break;
                default:
                    break;
            }
        }
        ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).set(null);
        super.channelInactive(ctx);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        logger.trace("exceptionCaught: {}", ctx, cause);
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, ProtoMessage msg) throws Exception {
        switch (msg.getType()) {
            case HEARTBEAT_PING:
                handlePingMessage(ctx, msg);
                break;
            case REQUEST:
                handleRequestMessage(ctx, msg);
                break;
            case TRANSFER:
                handleTransferMessage(ctx, msg);
                break;
            case LOCAL_CONNECTED:
                handleLocalConnectedMessage(ctx, msg);
                break;
            case LOCAL_DISCONNECT:
                handleLocalDisconnectMessage(ctx, msg);
                break;
            default:
                break;
        }
    }

    /**
     * 处理Ping消息
     */
    private void handlePingMessage(ChannelHandlerContext ctx, ProtoMessage msg) throws Exception {
        ctx.writeAndFlush(new ProtoMessage(ProtoMessage.Type.HEARTBEAT_PONG, null, null));
    }

    /**
     * 处理请求开启隧道消息
     */
    private void handleRequestMessage(ChannelHandlerContext ctx, ProtoMessage msg) throws Exception {
        ProtoRequest protoRequest = ProtoRequest.fromBytes(msg.getHead());
        switch (protoRequest.proto()) {
            case TCP:
                handleTcpRequestMessage(ctx, tcpTunnelServer, protoRequest);
                break;
            case HTTP:
                if (httpTunnelServer != null) {
                    handleHttpRequestMessage(ctx, httpTunnelServer, protoRequest);
                }
                break;
            case HTTPS:
                if (httpsTunnelServer != null) {
                    handleHttpRequestMessage(ctx, httpsTunnelServer, protoRequest);
                }
                break;
            default:
                final ByteBuf head = Unpooled.copyBoolean(false);
                ctx.channel().writeAndFlush(
                    new ProtoMessage(
                        ProtoMessage.Type.RESPONSE,
                        head.array(),
                        "协议错误".getBytes(StandardCharsets.UTF_8)
                    )
                ).addListener(ChannelFutureListener.CLOSE);
                head.release();
                break;
        }
    }


    /**
     * 处理数据传输消息
     */
    private void handleTransferMessage(ChannelHandlerContext ctx, ProtoMessage msg) {
        final ServerTunnelSessions tunnelSessions = ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).get();
        if (tunnelSessions != null) {
            final ByteBuf head = Unpooled.wrappedBuffer(msg.getHead());
            final long tunnelToken = head.readLong();
            final long sessionToken = head.readLong();
            head.release();
            final Proto proto = tunnelSessions.protoRequest().proto();
            switch (proto) {
                case TCP:
                    final Channel tcpSessionChannel = tcpTunnelServer.getSessionChannel(tunnelToken, sessionToken);
                    if (tcpSessionChannel != null) {
                        tcpSessionChannel.writeAndFlush(Unpooled.wrappedBuffer(msg.getData()));
                    }
                    break;
                case HTTP:
                    if (httpTunnelServer == null) {
                        break;
                    }
                    final Channel httpSessionChannel = httpTunnelServer.getSessionChannel(tunnelToken, sessionToken);
                    if (httpSessionChannel != null) {
                        httpSessionChannel.writeAndFlush(Unpooled.wrappedBuffer(msg.getData()));
                    }
                    break;
                case HTTPS:
                    if (httpsTunnelServer == null) {
                        break;
                    }
                    final Channel httpsSessionChannel = httpsTunnelServer.getSessionChannel(tunnelToken, sessionToken);
                    if (httpsSessionChannel != null) {
                        httpsSessionChannel.writeAndFlush(Unpooled.wrappedBuffer(msg.getData()));
                    }
                    break;
                default:
                    break;
            }
        }
    }

    /**
     * 处理本地连接成功消息
     */
    private void handleLocalConnectedMessage(ChannelHandlerContext ctx, ProtoMessage msg) {
        logger.trace("LocalConnected: {}", ctx);
        // 无须处理
    }

    /**
     * 处理本地连接断开消息
     */
    private void handleLocalDisconnectMessage(ChannelHandlerContext ctx, ProtoMessage msg) {
        logger.trace("LocalDisconnect: {}", ctx);
        final ByteBuf head = Unpooled.wrappedBuffer(msg.getHead());
        final long tunnelToken = head.readLong();
        final long sessionToken = head.readLong();
        head.release();

        ServerTunnelSessions tunnelSessions = ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).get();
        if (tunnelSessions == null) {
            return;
        }
        final Channel sessionChannel = tunnelSessions.getSessionChannel(sessionToken);
        if (sessionChannel != null) {
            // 解决 HTTP/1.x 数据传输问题
            sessionChannel.writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
        }
    }

    private void handleTcpRequestMessage(
        @NotNull ChannelHandlerContext ctx,
        @NotNull TcpTunnelServer tcpTunnelServer,
        @NotNull ProtoRequest protoRequest
    ) throws Exception {
        try {
            protoRequest = protoRequestInterceptor.handleProtoRequest(protoRequest);
            final long tunnelToken = tunnelTokenProducer.nextToken();
            final ServerTunnelSessions tunnelSessions = new ServerTunnelSessions(tunnelToken, protoRequest, ctx.channel());

            ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).set(tunnelSessions);

            tcpTunnelServer.startTunnel(null, protoRequest.remotePort(), tunnelSessions);

            final ByteBuf head = Unpooled.buffer(9);
            head.writeBoolean(true);
            head.writeLong(tunnelToken);
            ctx.channel().writeAndFlush(
                new ProtoMessage(
                    ProtoMessage.Type.RESPONSE,
                    head.array(),
                    protoRequest.toBytes()
                )
            );
            head.release();

        } catch (ProtoException e) {
            final ByteBuf head = Unpooled.copyBoolean(false);
            ctx.channel().writeAndFlush(
                new ProtoMessage(
                    ProtoMessage.Type.RESPONSE,
                    head.array(),
                    e.getMessage().getBytes(StandardCharsets.UTF_8)
                )
            ).addListener(ChannelFutureListener.CLOSE);
            head.release();
        }
    }

    private void handleHttpRequestMessage(
        @NotNull ChannelHandlerContext ctx,
        @NotNull HttpTunnelServer server,
        @NotNull ProtoRequest protoRequest
    ) throws Exception {
        try {
            protoRequest = protoRequestInterceptor.handleProtoRequest(protoRequest);
            if (server.isRegistered(protoRequest.vhost())) {
                throw new ProtoException("vhost(" + protoRequest.vhost() + ") already used");
            }
            final long tunnelToken = tunnelTokenProducer.nextToken();
            final ServerTunnelSessions tunnelSessions = new ServerTunnelSessions(tunnelToken, protoRequest, ctx.channel());
            ctx.channel().attr(AttributeKeys.SERVER_TUNNEL_SESSIONS).set(tunnelSessions);
            server.register(protoRequest.vhost(), tunnelSessions);
            final ByteBuf head = Unpooled.buffer(9);
            head.writeBoolean(true);
            head.writeLong(tunnelToken);
            ctx.channel().writeAndFlush(
                new ProtoMessage(
                    ProtoMessage.Type.RESPONSE,
                    head.array(),
                    protoRequest.toBytes()
                )
            );
            head.release();
        } catch (ProtoException e) {
            final ByteBuf head = Unpooled.copyBoolean(false);
            ctx.channel().writeAndFlush(
                new ProtoMessage(
                    ProtoMessage.Type.RESPONSE,
                    head.array(),
                    e.getMessage().getBytes(StandardCharsets.UTF_8)
                )
            ).addListener(ChannelFutureListener.CLOSE);
            head.release();
        }
    }

}

File: tunnel2-server/src/main/kotlin/tunnel2/server/TunnelServerChannelHandler.kt
package tunnel2.server

import io.netty.buffer.Unpooled
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import tunnel2.common.TunnelException
import tunnel2.common.TunnelRequest
import tunnel2.common.TunnelType
import tunnel2.common.logging.LoggerFactory
import tunnel2.common.proto.ProtoCw
import tunnel2.common.proto.ProtoMessage
import tunnel2.server.http.HttpServer
import tunnel2.server.interceptor.TunnelRequestInterceptor
import tunnel2.server.internal.AK_SERVER_SESSION_CHANNELS
import tunnel2.server.internal.IdProducer
import tunnel2.server.internal.ServerSessionChannels
import tunnel2.server.tcp.TcpServer

class TunnelServerChannelHandler(
    private val tunnelRequestInterceptor: TunnelRequestInterceptor,
    private val tunnelIdProducer: IdProducer,
    private val tcpServer: TcpServer? = null,
    private val httpServer: HttpServer? = null,
    private val httpsServer: HttpServer? = null
) : SimpleChannelInboundHandler<ProtoMessage>() {

    companion object {
        private val logger = LoggerFactory.getLogger(TunnelServerChannelHandler::class.java)
    }

    override fun channelInactive(ctx: ChannelHandlerContext?) {
        logger.trace("channelInactive: {}", ctx)
        if (ctx == null) {
            super.channelInactive(ctx)
            return
        }
        ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).get()?.also {
            when (it.tunnelRequest.type) {
                TunnelType.TCP -> tcpServer?.registry?.unregister(it.tunnelId)
                TunnelType.HTTP -> httpServer?.registry?.unregister(it.tunnelRequest.host)
                TunnelType.HTTPS -> httpsServer?.registry?.unregister(it.tunnelRequest.host)
                else -> {
                    // Nothing
                }
            }
        }
        ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).set(null)
        super.channelInactive(ctx)
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext?, cause: Throwable?) {
        ctx ?: return
        logger.trace("exceptionCaught: {}", ctx, cause)
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
    }

    override fun channelRead0(ctx: ChannelHandlerContext?, msg: ProtoMessage?) {
        ctx ?: return
        msg ?: return
        when (msg.cw) {
            ProtoCw.PING -> handlePingMessage(ctx, msg)
            ProtoCw.REQUEST -> handleRequestMessage(ctx, msg)
            ProtoCw.TRANSFER -> handleTransferMessage(ctx, msg)
            ProtoCw.LOCAL_CONNECTED -> handleLocalConnectedMessage(ctx, msg)
            ProtoCw.LOCAL_DISCONNECT -> handleLocalDisconnectMessage(ctx, msg)
            else -> {
                // Nothing
            }
        }
    }

    private fun handlePingMessage(ctx: ChannelHandlerContext, msg: ProtoMessage) {
        logger.trace("handlePingMessage# {}, {}", ctx, msg)
        ctx.writeAndFlush(ProtoMessage(ProtoCw.PONG))
    }

    private fun handleRequestMessage(ctx: ChannelHandlerContext, msg: ProtoMessage) {
        logger.trace("handleRequestMessage# {}, {}", ctx, msg)
        try {
            val tunnelRequest = TunnelRequest.fromBytes(msg.data)
            logger.trace("tunnelRequest: {}", tunnelRequest)
            when (tunnelRequest.type) {
                TunnelType.TCP -> {
                    tcpServer?.also { handleTcpRequestMessage(ctx, it, tunnelRequest) }
                        ?: throw TunnelException("TCP协议隧道未开启")
                }
                TunnelType.HTTP -> {
                    httpServer?.also { handleHttpRequestMessage(ctx, it, tunnelRequest) }
                        ?: throw TunnelException("HTTP协议隧道未开启")
                }
                TunnelType.HTTPS -> {
                    httpsServer?.also { handleHttpRequestMessage(ctx, it, tunnelRequest) }
                        ?: throw TunnelException("HTTPS协议隧道未开启")
                }
                else -> throw TunnelException("不支持的隧道类型")
            }
        } catch (e: Exception) {
            ctx.channel().writeAndFlush(
                ProtoMessage(
                    ProtoCw.RESPONSE_ERR,
                    data = e.message.toString().toByteArray()
                )
            ).addListener(ChannelFutureListener.CLOSE)
        }
    }

    private fun handleTransferMessage(ctx: ChannelHandlerContext, msg: ProtoMessage) {
        val sessionChannels = ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).get() ?: return
        when (sessionChannels.tunnelRequest.type) {
            TunnelType.TCP -> {
                tcpServer ?: return
                tcpServer.registry.getSessionChannel(msg.tunnelId, msg.sessionId)?.writeAndFlush(
                    Unpooled.wrappedBuffer(msg.data)
                )
            }
            TunnelType.HTTP -> {
                httpServer ?: return
                httpServer.registry.getSessionChannel(msg.tunnelId, msg.sessionId)?.writeAndFlush(
                    Unpooled.wrappedBuffer(msg.data)
                )
            }
            TunnelType.HTTPS -> {
                httpsServer ?: return
                httpsServer.registry.getSessionChannel(msg.tunnelId, msg.sessionId)?.writeAndFlush(
                    Unpooled.wrappedBuffer(msg.data)
                )
            }
            else -> {
                // pass
            }
        }
    }

    private fun handleLocalConnectedMessage(ctx: ChannelHandlerContext, msg: ProtoMessage) {
        logger.trace("handleLocalConnectedMessage# {}, {}", ctx, msg)
        // 无须处理
    }

    private fun handleLocalDisconnectMessage(ctx: ChannelHandlerContext, msg: ProtoMessage) {
        logger.trace("handleLocalDisconnectMessage# {}, {}", ctx, msg)
        val sessionChannels = ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).get() ?: return
        val sessionChannel = sessionChannels.getSessionChannel(msg.sessionId)
        // 解决 HTTP/1.x 数据传输问题
        sessionChannel?.writeAndFlush(Unpooled.EMPTY_BUFFER)?.addListener(ChannelFutureListener.CLOSE)
    }

    // ==================================================================
    @Throws(TunnelException::class)
    private fun handleTcpRequestMessage(ctx: ChannelHandlerContext, server: TcpServer, request: TunnelRequest) {
        val tunnelRequest = tunnelRequestInterceptor.handleTunnelRequest(request)
        val tunnelId = tunnelIdProducer.nextId
        val sessionChannels = ServerSessionChannels(
            tunnelId,
            tunnelRequest,
            ctx.channel()
        )
        ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).set(sessionChannels)
        server.startTunnel(null, tunnelRequest.remotePort, sessionChannels)
        ctx.channel().writeAndFlush(
            ProtoMessage(
                ProtoCw.RESPONSE_OK,
                tunnelId = tunnelId,
                data = tunnelRequest.toBytes()
            )
        )
    }

    @Throws(TunnelException::class)
    private fun handleHttpRequestMessage(ctx: ChannelHandlerContext, server: HttpServer, request: TunnelRequest) {
        val tunnelRequest = tunnelRequestInterceptor.handleTunnelRequest(request)
        if (server.registry.isRegistered(tunnelRequest.host)) {
            throw TunnelException("host(${tunnelRequest.host}) already used")
        }
        val tunnelId = tunnelIdProducer.nextId
        val sessionChannels = ServerSessionChannels(tunnelId, tunnelRequest, ctx.channel())
        ctx.channel().attr<ServerSessionChannels>(AK_SERVER_SESSION_CHANNELS).set(sessionChannels)
        server.registry.register(tunnelRequest.host, sessionChannels)
        ctx.channel().writeAndFlush(
            ProtoMessage(
                ProtoCw.RESPONSE_OK,
                tunnelId = tunnelId,
                data = tunnelRequest.toBytes()
            )
        )
    }

}